### PR TITLE
feature(iot-service): Add support for metadata properties with applied configuration

### PR DESCRIPTION
**Pull request by bikamani**
When a configuration is applied on Twin, there are two metadata properties which we currently we do not deserialize.

image

This PR adds in following

- Support for additional metadata properties: lastUpdatedBy, lastUpdatedByDigest
- Refactor tests

Validated changes with sample, below is printing twin with additional metadata properties
`Getting the Device twin
Device Id: bikamani_dev
ETag: AAAAAAAAABQ=
Version: 21
Model Id: 
Tags: {"HomeID":"7fe1bf5e-60bf-42ea-839a-40c507f9f0fe"}
Reported properties: {}
Desired properties: {"hum":90.0,"temp":75.0,"chiller-water":{"temperature":66.0,"pressure":28.0},"$version":20,"$metadata":{"$lastUpdated":"2021-05-20T21:05:36.195Z","$lastUpdatedVersion":20,"hum":{"$lastUpdated":"2021-05-20T21:05:36.195Z","$lastUpdatedVersion":20,"$**lastUpdatedBy**":"newconfig","$**lastUpdatedByDigest**":"637570574076206429"},"temp":{"$lastUpdated":"2021-05-20T21:05:15.148Z","$lastUpdatedVersion":19},"chiller-water":{"$lastUpdated":"2021-05-20T21:05:36.195Z","$lastUpdatedVersion":20,"temperature":{"$lastUpdated":"2021-05-20T21:05:36.195Z","$lastUpdatedVersion":20,"$lastUpdatedBy":"test","$lastUpdatedByDigest":"637570515479675333"},"pressure":{"$lastUpdated":"2021-05-20T21:05:36.195Z","$lastUpdatedVersion":20,"$lastUpdatedBy":"test","$lastUpdatedByDigest":"637570515479675333"}}}}`

### CODE REVIEW

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java [hunk 1]
```diff
@@ -164,14 +187,43 @@ else if(key.equals(LAST_UPDATE_VERSION_TAG))
...
+            else if (key.equals(TwinMetadata.LAST_UPDATED_BY))
+            {
+                lastUpdatedBy = (String)entry.getValue();
+            }
+            else if (key.equals(TwinMetadata.LAST_UPDATED_BY_DIGEST))
+            {
+                lastUpdatedByDigest = (String)entry.getValue();
+            }
         }
+        // TODO: additional checks
```
Comment (bikamani) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java:199]: awaiting response from service team to understand if we need to check for null values in newly added properties.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import com.microsoft.azure.sdk.iot.deps.serializer.ParserUtility;
 import com.microsoft.azure.sdk.iot.deps.util.Tools;
 
+import javax.swing.*;
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java:11]: Unused import?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java [hunk 3]
```diff
@@ -66,10 +66,12 @@
  * </pre>
  *
  * <p> This class exposes the Twin collection with or without metadata as a Map here
- *     user can gat both the value and the metadata. For instance, in the above TwinCollection,
+ *     user can get both the value and the metadata. For instance, in the above TwinCollection,
  *     {@link #get(Object)} for <b>Color</b> will return <b>White</b> and the {@link #getTwinMetadataFinal(String)}
  *     for <b>Color</b> will return the Object TwinMetadata that contain {@link TwinMetadata#getLastUpdated()}
- *     that will returns the {@code Date} <b>2017-09-21T02:07:44.238Z</b> and {@link TwinMetadata#getLastUpdatedVersion()}
+ *     that will returns the {@code Date} <b>2017-09-21T02:07:44.238Z</b>, {@link TwinMetadata#getLastUpdatedBy()} ()}
+ *     that will returns the {@code String}, {@link TwinMetadata#getLastUpdatedByDigest()} ()}
```
Comment (timtay-microsoft) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:73]: Unnecessary "()" within these brackets. Here and the line above

**Review by timtay-microsoft — APPROVED** [review]

**Review by abhipsaMisra — APPROVED** [review]

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java [hunk 4]
```diff
@@ -66,10 +66,12 @@
  * </pre>
  *
  * <p> This class exposes the Twin collection with or without metadata as a Map here
- *     user can gat both the value and the metadata. For instance, in the above TwinCollection,
+ *     user can get both the value and the metadata. For instance, in the above TwinCollection,
  *     {@link #get(Object)} for <b>Color</b> will return <b>White</b> and the {@link #getTwinMetadataFinal(String)}
  *     for <b>Color</b> will return the Object TwinMetadata that contain {@link TwinMetadata#getLastUpdated()}
- *     that will returns the {@code Date} <b>2017-09-21T02:07:44.238Z</b> and {@link TwinMetadata#getLastUpdatedVersion()}
+ *     that will returns the {@code Date} <b>2017-09-21T02:07:44.238Z</b>, {@link TwinMetadata#getLastUpdatedBy()}
```
Comment (drwill-ms) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:72]: "that will returns"  should either be "that returns" or "that will return"

Comment (drwill-ms) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:72]: Why do we have a date example and word it like we'll return this exact date. Are we trying to say we'll return a date in this format?

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java [hunk 5]
```diff
@@ -477,6 +489,14 @@ private void fillJsonMetadata(JsonObject jsonMetadata)
         {
             jsonMetadata.addProperty(TwinMetadata.LAST_UPDATE_TAG, ParserUtility.dateTimeUtcToString(this.twinMetadata.getLastUpdated()));
             jsonMetadata.addProperty(TwinMetadata.LAST_UPDATE_VERSION_TAG, this.twinMetadata.getLastUpdatedVersion());
+            if(this.twinMetadata.getLastUpdatedBy() != null)
```
Comment (drwill-ms) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:492]: space after `if`

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java (hunk 4 above)
Comment (bikamani) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:72]: I read it as "return date and for example 2017-09-21(date format)". What do you think as a better way to describe it?

Comment (bikamani) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:72]: I have added "for example" before this value, also added examples for the fields I added.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java (hunk 5 above)
Comment (bikamani) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:493]: I did add a space here to keep it consistent with other places.

Comment (drwill-ms) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:493]: I don't see that a space got added.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java [hunk 6]
```diff
@@ -477,6 +490,14 @@ private void fillJsonMetadata(JsonObject jsonMetadata)
...
+            {
+                jsonMetadata.addProperty(TwinMetadata.LAST_UPDATED_BY, this.twinMetadata.getLastUpdatedBy());
+            }
+            if(this.twinMetadata.getLastUpdatedByDigest() != null)
```
Comment (drwill-ms) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinCollection.java:497]: same here, `if (`

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java [hunk 7]
```diff
@@ -164,14 +188,43 @@ else if(key.equals(LAST_UPDATE_VERSION_TAG))
...
         if((lastUpdatedVersion != null) || !Tools.isNullOrEmpty(lastUpdated))
         {
-            return new TwinMetadata(lastUpdated, lastUpdatedVersion);
+            return new TwinMetadata(lastUpdated, lastUpdatedVersion, lastUpdatedBy, lastUpdatedByDigest);
         }
         return null;
     }
 
+    /**
+     * Getter for the LastUpdatedBy.
```
Comment (drwill-ms) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java:209]: Leave out the word "the", so it is: Getter for LastUpdatedBy.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java [hunk 8]
```diff
@@ -164,14 +188,43 @@ else if(key.equals(LAST_UPDATE_VERSION_TAG))
...
+     *
+     * @return the {@code String} representing the configuration LastUpdatedBy.
+     */
+    public String getLastUpdatedBy()
+    {
+        return this.lastUpdatedBy;
+    }
+
+    /**
+     * Getter for the lastUpdatedByDigest.
```
Comment (drwill-ms) [inline deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java:219]: Line 209 uses PascalCase for the object, but this one uses camelCase

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java (hunk 7 above)
Comment (bikamani) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java:209]: sure let me address this in a follow up PR.

File: deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java (hunk 8 above)
Comment (bikamani) [reply deps/src/main/java/com/microsoft/azure/sdk/iot/deps/twin/TwinMetadata.java:219]: will address in follow up PR.

**Pull request merged**